### PR TITLE
Adding  some audio transforms and augmentations to tonic

### PR DESCRIPTION
**Pull request by MinaKh**
This branch includes following updates in tonic/develop:

-  Three transforms are added to `audio_transforms.py`: 
     - `SwapAxes`, `AmplitudeScale` and  `robustAmplitudeScale`

-  A new script is added: `audio_augmentations.py` containing wrapper classes for following audio augmentations:
    -  `RandomAmplitudeScale`
    - `RandomPitchShift`
    -  `RandomTimeStreatch`
    -  `RIR`: adding room impulse response (echo effect) 
    - `AddWhiteNoise`

- corresponding tests added: `test_audio_trnasform.py`
- A new script added for testing audio augmentations  ---> `test_audio_augmentations.py`
- A tutorial notebook is added in `docs/tutorials/ `

### CODE REVIEW

**Comment by fabrizio-ottati** [comment]
@MinaKh I think you should add `torchaudio` to the test `requirements.txt`, 'cause it fails as dependency in the CI tests.

**Comment by MinaKh** [comment]
> @MinaKh I think you should add `torchaudio` to the test `requirements.txt`, 'cause it fails as dependency in the CI tests.

Thanks, just did!

**Comment by fabrizio-ottati** [comment]
It seems that the tests now are running using CUDA. Maybe it is better to run them with everything on CPU? Or we could modify the `requirements.txt` to install the GPU version of PyTorch (both `vision` and `audio`).

**Comment by MinaKh** [comment]
> It seems that the tests now are running using CUDA. Maybe it is better to run them with everything on CPU? Or we could modify the `requirements.txt` to install the GPU version of PyTorch (both `vision` and `audio`).

@fabrizio-ottati No, I can set them to run on cpu. Also there are other tests that I passed on my machine but fail on CI. I need to fix without need to install extra packages. Thanks!

File: tonic/audio_augmentations.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+import random
+from dataclasses import dataclass, field
+from typing import Optional
+
+import librosa
+import numpy as np
+import torch
+import torchaudio
+import torchaudio.functional as F
+from qut_noise import QUTNoise
```
Comment (biphasic) [inline tonic/audio_augmentations.py:11]: this is a dependency that I won't be able to add to this library like that.

File: tonic/audio_augmentations.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+            self.aug_index = random.choice(range(0, len(self.factors)))
+
+        noise = np.random.normal(0, audio.std(), audio.size)
+        noise_factor = self.factors[self.aug_index]
+        noisy_audio = audio + noise_factor * noise
+        return noisy_audio
+
+
+@dataclass
+class AddHomeNoise:
```
Comment (biphasic) [inline tonic/audio_augmentations.py:178]: This is very specific, is there a way how you can abstract it to noise from different datasets?

File: tonic/audio_augmentations.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    Returns:
+        np.ndarray: data sample convolved with RIR
+    """
+
+    samplerate: float
+    caching: bool = False
+
+    def __call__(self, audio):
+        SAMPLE_RIR = download_asset(
+            "tutorial-assets/Lab41-SRI-VOiCES-rm1-impulse-mc01-stu-clo-8000hz.wav"
```
Comment (biphasic) [inline tonic/audio_augmentations.py:388]: this sort of hardcoded things I cannot merge into a public library

**Review by biphasic — CHANGES_REQUESTED** [review]
Most transforms are encapsulated enough so that I can add them, but some stuff that uses QUTNoise things I won't be able to merge like that, unless it is made a bit more general. For example, maybe Tonic has a AddNoise class, but then in your user code at SynSense you call it with AddNoise(QUTNoise), after the principle of dependency injection

**Comment by biphasic** [comment]
The Github Actions test suite doens't have a GPU installed, therefore there's no point in installing any CUDA dependencies. Tests should always run on CPU please. Thank you

**Comment by MinaKh** [comment]
> The Github Actions test suite doens't have a GPU installed, therefore there's no point in installing any CUDA dependencies. Tests should always run on CPU please. Thank you

Thanks @biphasic, As far as I checked my tests are not running on GPU. The error might be caused by general imports of torch and torchaudio, or by the difference in my local version and the installed one on the server. So I included the versions in the requirements. 
At this point I need to be able to run tests on GitHub to understand the issue better and fix it. Currently tests are not running automatically after my pushes (perhaps needs to be authorized every time by you and other admins?).

File: requirements.txt [hunk 4]
```diff
@@ -7,3 +7,5 @@ typing_extensions
 librosa
 pbr
 expelliarmus
+torch==1.12.0+cu113
+torchaudio==0.12.0+cu113
```
Comment (biphasic) [inline requirements.txt:11]: cu113 means CUDA 11.3, so this is installing a fixed version of pytorch with CUDA backend

**Comment by biphasic** [comment]
@MinaKh are the tests passing on your local machine?

**Comment by biphasic** [comment]
Also I just relaxed the Github actions approval to the minimum level possible, I hope it now works without my manual approval!

**Comment by MinaKh** [comment]
Hi @biphasic, 
Thanks for facilitating this! It seems that we need to explicitly install cpu version of `torchaudio` in `ci-pipeline.yml`. 
Lines 24, 45 and 66 need to be updated to following:
`  pip install torch torchaudio torchvision --index-url https://download.pytorch.org/whl/cpu
`
I tried to do that but apparently I don't have permission to update the workflow.

**Comment by biphasic** [comment]
where do you see that error? I see a different error, again related to CUDA
https://github.com/neuromorphs/tonic/actions/runs/7089430652/job/19294156537?pr=273

**Comment by MinaKh** [comment]
> where do you see that error? I see a different error, again related to CUDA https://github.com/neuromorphs/tonic/actions/runs/7089430652/job/19294156537?pr=273

**Comment by MinaKh** [comment]
> where do you see that error? I see a different error, again related to CUDA https://github.com/neuromorphs/tonic/actions/runs/7089430652/job/19294156537?pr=273

@biphasic Yes I see this error and I think what causes that is we dont specify the cpu version while installing torchaudio. Tofix that I proposed that update in the `ci-pipeline.yml` 
And the error that I included in  the last message is what I see in my pc. not being able to push to `ci-pipeline.yml` . I am not authorized to push to workflow

**Comment by fabrizio-ottati** [comment]
Will get on it! On 5 Dec 2023, at 10:54, Mina Khoei ***@***.***> wrote:﻿

where do you see that error? I see a different error, again related to CUDA https://github.com/neuromorphs/tonic/actions/runs/7089430652/job/19294156537?pr=273

@biphasic Yes I see this error and I think what causes that is we dont specify the cpu version while installing torchaudio. Tofix that I proposed that update in the ci-pipeline.yml
And the error that I included in  the last message is what I see in my pc. not being able to push to ci-pipeline.yml . I am not authorized to push to workflow

**Comment by MinaKh** [comment]
@fabrizio-ottati Thanks Fabrizio, sorry I just accidentally closed this PR :D Can you undo that?

**Comment by MinaKh** [comment]
Ok, reopened! sorry for the inconvenience!

File: test/requirements.txt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 pytest
 coverage
 torch
+torchaudio
```
Comment (fabrizio-ottati) [inline test/requirements.txt:4]: @MinaKh here's where you have to specify the CPU version of `torchaudio`. The CI pipeline installs the dependencies from this file.

**Comment by fabrizio-ottati** [comment]
@MinaKh I need to sit and do it properly. I will update you this afternoon

**Comment by fabrizio-ottati** [comment]
image

I don't know why but it keeps installing the CUDA version even if I have specificied to use the CPU wheel of PyTorch.

**Comment by fabrizio-ottati** [comment]
Okay, it seems I convinced it @MinaKh :)

**Comment by MinaKh** [comment]
> Okay, it seems I convinced it @MinaKh :)



> Okay, it seems I convinced it @MinaKh :)

Thanks @fabrizio-ottati

**Comment by MinaKh** [comment]
> Most transforms are encapsulated enough so that I can add them, but some stuff that uses QUTNoise things I won't be able to merge like that, unless it is made a bit more general. For example, maybe Tonic has a AddNoise class, but then in your user code at SynSense you call it with AddNoise(QUTNoise), after the principle of dependency injection

@biphasic I removed those classes (noise augmentations) from this branch. currently it is very specific and I will prepare another PR later for that.

**Comment by fabrizio-ottati** [comment]
What's the status on this PR? :)

File: tonic/audio_augmentations.py (hunk 2 above)
Comment (MinaKh) [reply tonic/audio_augmentations.py]: @biphasic I have removed all noise related augmentations from this PR.

**Comment by MinaKh** [comment]
> What's the status on this PR? :)

It is ready for final review...

**Review by biphasic — APPROVED** [review]

**Pull request merged**